File: sections/27_bitmask_dp/problems/b_assignment_profit/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    profit = [[next(it) for _ in range(n)] for _ in range(n)]
    total = 1 << n
    neg = -10**30
    dp = [neg] * total
    dp[0] = 0

    for mask in range(total):
        worker = mask.bit_count()
        if worker == n or dp[mask] == neg:
            continue
        for job in range(n):
            if not (mask >> job) & 1:
                nxt = mask | (1 << job)
                value = dp[mask] + profit[worker][job]
                if value > dp[nxt]:
                    dp[nxt] = value

    print(dp[total - 1])


if __name__ == "__main__":
    main()

File: sections/27_bitmask_dp/problems/b_assignment_profit/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<vector<long long>> profit(n, vector<long long>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) cin >> profit[i][j];
    }

    int total = 1 << n;
    const long long NEG = -(1LL << 60);
    vector<long long> dp(total, NEG);
    dp[0] = 0;

    for (int mask = 0; mask < total; mask++) {
        int worker = __builtin_popcount((unsigned)mask);
        if (worker == n || dp[mask] == NEG) continue;
        for (int job = 0; job < n; job++) {
            if ((mask & (1 << job)) == 0) {
                int nxt = mask | (1 << job);
                dp[nxt] = max(dp[nxt], dp[mask] + profit[worker][job]);
            }
        }
    }

    cout << dp[total - 1] << '\n';
    return 0;
}
